File: sdk/node-ts/lib/egress-intercept.ts
/**
 * High-level egress interception API with Gondolin/mitmproxy-style hooks.
 *
 * Wraps the low-level `EgressStream` (from native Rust binding) with
 * `onRequest`/`onResponse` callback hooks.
 *
 * @example
 * ```ts
 * import { Sandbox } from 'microsandbox';
 * import { egressIntercept } from 'microsandbox/egress-intercept';
 *
 * const handle = await egressIntercept(sandbox, {
 *   onRequest: async (request, ctx) => {
 *     console.log(`→ ${request.method} ${request.uri} [${ctx.sni}]`);
 *     request.headers.push(["X-Trace-Id", "abc123"]);
 *     return request; // forward modified
 *   },
 *   onResponse: async (response, request, ctx) => {
 *     console.log(`← ${response.status} [${ctx.sni}]`);
 *     return undefined; // pass through
 *   },
 * });
 *
 * // ... do other work while interception is active ...
 *
 * await handle.stop();  // graceful shutdown
 * ```
 */

import type { EgressEvent, EgressHttpRequest, EgressHttpResponse } from "../index.d.cts";

/**
 * Connection metadata extracted from an {@link EgressEvent}.
 *
 * This is a convenience type — not a duplicate of a Rust binding type.
 * It groups the per-event metadata fields so hook signatures stay clean.
 */
export interface EgressContext {
  sni: string;
  dst: string;
  connectionId: number;
  timestampMs: number;
}

export interface EgressInterceptOptions {
  /**
   * Called for each outbound HTTP request.
   *
   * Return values:
   * - `undefined` → pass through unchanged
   * - `EgressHttpRequest` → forward modified request to server
   * - `EgressHttpResponse` → short-circuit (return response to guest, skip server)
   * - throw → block the connection
   */
  onRequest?: (
    request: EgressHttpRequest,
    ctx: EgressContext
  ) => Promise<EgressHttpRequest | EgressHttpResponse | undefined> | EgressHttpRequest | EgressHttpResponse | undefined;

  /**
   * Called for each server response.
   *
   * Return values:
   * - `undefined` → pass through unchanged
   * - `EgressHttpResponse` → forward modified response to guest
   * - throw → block the connection
   */
  onResponse?: (
    response: EgressHttpResponse,
    request: EgressHttpRequest | undefined,
    ctx: EgressContext
  ) => Promise<EgressHttpResponse | undefined> | EgressHttpResponse | undefined;
}

/**
 * Handle for a running egress interception loop.
 *
 * Returned by {@link egressIntercept}. The event loop runs in the
 * background. Use {@link stop} for graceful shutdown or await {@link done}
 * to wait for natural completion (e.g., sandbox stopped).
 */
export class EgressInterceptHandle {
  private _done: Promise<void>;
  private _stopped = false;

  /** @internal */
  constructor(donePromise: Promise<void>) {
    this._done = donePromise;
  }

  /** Promise that resolves when the intercept loop ends (sandbox stopped or {@link stop} called). */
  get done(): Promise<void> {
    return this._done;
  }

  /** Signal the intercept loop to stop after the current event. */
  stop(): Promise<void> {
    this._stopped = true;
    return this._done;
  }

  /** @internal */
  get stopped(): boolean {
    return this._stopped;
  }
}

/**
 * Check if a return value is a response (has `status` field) vs a request (has `method` field).
 */
function isResponse(value: any): value is EgressHttpResponse {
  return value && typeof value.status === "number";
}

/**
 * Start egress interception with callback hooks.
 *
 * Connects to the sandbox's egress socket and starts processing events
 * in the background. Returns an {@link EgressInterceptHandle} immediately.
 *
 * @param sandbox - A running `Sandbox` instance with `egressInterceptHosts` configured.
 * @param options - `onRequest` and/or `onResponse` callback hooks.
 * @returns A handle to control the interception loop.
 */
export async function egressIntercept(
  sandbox: any, // Sandbox type from native binding
  options: EgressInterceptOptions
): Promise<EgressInterceptHandle> {
  const stream = await sandbox.egressConnection();
  const lastRequests = new Map<number, EgressHttpRequest>();

  let resolveLoop!: () => void;
  const donePromise = new Promise<void>((resolve) => { resolveLoop = resolve; });
  const handle = new EgressInterceptHandle(donePromise);

  (async () => {
    try {
      while (!handle.stopped) {
        const event: EgressEvent | null = await stream.recv();
        if (event === null) break;

        const ctx: EgressContext = {
          sni: event.sni,
          dst: event.dst,
          connectionId: event.connectionId,
          timestampMs: event.timestampMs,
        };

        try {
          if (event.kind === "request" && event.request) {
            const request: EgressHttpRequest = event.request;
            lastRequests.set(event.connectionId, request);

            if (options.onRequest) {
              const result = await options.onRequest(request, ctx);

              if (result === undefined) {
                await stream.passThrough(event.id);
              } else if (isResponse(result)) {
                await stream.shortCircuit(event.id, result);
              } else {
                // Modified request
                lastRequests.set(event.connectionId, result);
                await stream.modifyRequest(event.id, result);
              }
            } else {
              await stream.passThrough(event.id);
            }
          } else if (event.kind === "response" && event.response) {
            const response: EgressHttpResponse = event.response;
            const originalRequest = lastRequests.get(event.connectionId);
            lastRequests.delete(event.connectionId);

            if (options.onResponse) {
              const result = await options.onResponse(response, originalRequest, ctx);

              if (result === undefined) {
                await stream.passThrough(event.id);
              } else {
                await stream.modifyResponse(event.id, result);
              }
            } else {
              await stream.passThrough(event.id);
            }
          } else {
            await stream.passThrough(event.id);
          }
        } catch {
          // Hook threw — block the connection
          await stream.block(event.id);
        }
      }
    } finally {
      resolveLoop();
    }
  })();

  return handle;
}

File: sdk/python/microsandbox/egress_intercept.py
"""High-level egress interception API with Gondolin/mitmproxy-style hooks.

Wraps the low-level ``EgressConnection`` (from native Rust binding) with
``on_request``/``on_response`` callback hooks.

Example::

    from microsandbox import Sandbox, Network
    from microsandbox.egress_intercept import egress_intercept
    from microsandbox.events import EgressHttpRequest, EgressHttpResponse

    sb = await Sandbox.create(
        name="agent", image="python",
        network=Network(egress_intercept_hosts=["*"]),
    )

    async def on_request(request, ctx):
        print(f"→ {request.method} {request.uri} [{ctx.sni}]")
        if "/admin" in request.uri:
            raise Exception("blocked")
        request.headers.append(("X-Trace-Id", "abc123"))
        return request

    async def on_response(response, request, ctx):
        print(f"← {response.status} [{ctx.sni}]")
        return None  # pass through

    handle = await egress_intercept(sb, on_request=on_request, on_response=on_response)
    # ... do other work while interception is active ...
    await handle.stop()   # graceful shutdown
"""

from __future__ import annotations

import asyncio
import contextlib
from collections.abc import Awaitable, Callable
from typing import Any

from microsandbox.events import (
    EgressContext,
    EgressHttpRequest,
    EgressHttpResponse,
)


class EgressInterceptHandle:
    """Handle for a running egress interception loop.

    Returned by :func:`egress_intercept`. The event loop runs in a
    background asyncio task.
    """

    def __init__(self, task: asyncio.Task[None]) -> None:
        self._task = task

    @property
    def done(self) -> asyncio.Task[None]:
        """Awaitable that resolves when the loop ends naturally (sandbox stopped)."""
        return self._task

    async def stop(self) -> None:
        """Cancel the intercept loop and wait for it to finish."""
        self._task.cancel()
        with contextlib.suppress(asyncio.CancelledError):
            await self._task


def _dict_to_request(d: dict[str, Any]) -> EgressHttpRequest:
    """Convert a raw event dict's request sub-dict to an EgressHttpRequest."""
    return EgressHttpRequest(
        method=d["method"],
        uri=d["uri"],
        headers=[(k, v) for k, v in d["headers"]],
        body=d.get("body"),
    )


def _dict_to_response(d: dict[str, Any]) -> EgressHttpResponse:
    """Convert a raw event dict's response sub-dict to an EgressHttpResponse."""
    return EgressHttpResponse(
        status=d["status"],
        headers=[(k, v) for k, v in d["headers"]],
        body=d.get("body"),
    )


def _is_response(value: Any) -> bool:
    """Check if a return value is a response (has ``status``) vs a request (has ``method``)."""
    return isinstance(value, EgressHttpResponse) or (
        isinstance(value, dict) and "status" in value
    )


async def egress_intercept(
    sandbox: Any,
    *,
    on_request: Callable[
        [EgressHttpRequest, EgressContext],
        Awaitable[EgressHttpRequest | EgressHttpResponse | None]
        | EgressHttpRequest
        | EgressHttpResponse
        | None,
    ] | None = None,
    on_response: Callable[
        [EgressHttpResponse, EgressHttpRequest | None, EgressContext],
        Awaitable[EgressHttpResponse | None] | EgressHttpResponse | None,
    ] | None = None,
) -> EgressInterceptHandle:
    """Start egress interception with callback hooks.

    Connects to the sandbox's egress socket and starts processing events
    in a background asyncio task. Returns an :class:`EgressInterceptHandle`
    immediately.

    Args:
        sandbox: A running ``Sandbox`` instance with ``egress_intercept_hosts`` configured.
        on_request: Called for each outbound HTTP request. Return values:

            - ``None`` → pass through unchanged
            - ``EgressHttpRequest`` → forward modified request to server
            - ``EgressHttpResponse`` → short-circuit (return response to guest)
            - raise any exception → block the connection

        on_response: Called for each server response. Return values:

            - ``None`` → pass through unchanged
            - ``EgressHttpResponse`` → forward modified response to guest
            - raise any exception → block the connection

    Returns:
        A handle to control the interception loop.
    """
    conn = await sandbox.egress_connection()

    async def _loop() -> None:
        last_requests: dict[int, EgressHttpRequest] = {}

        while True:
            event = await conn.recv()
            if event is None:
                break

            ctx = EgressContext(
                sni=event["sni"],
                dst=event["dst"],
                connection_id=event["connection_id"],
                timestamp_ms=event["timestamp_ms"],
            )
            event_id: int = event["id"]
            connection_id: int = event["connection_id"]

            try:
                if event["kind"] == "request" and "request" in event:
                    request = _dict_to_request(event["request"])
                    last_requests[connection_id] = request

                    if on_request is not None:
                        result = on_request(request, ctx)
                        if asyncio.iscoroutine(result):
                            result = await result

                        if result is None:
                            await conn.pass_through(event_id)
                        elif _is_response(result):
                            await conn.short_circuit(
                                event_id, result.status,
                                result.headers, result.body,
                            )
                        else:
                            last_requests[connection_id] = result
                            await conn.modify_request(
                                event_id, result.method, result.uri,
                                result.headers, result.body,
                            )
                    else:
                        await conn.pass_through(event_id)

                elif event["kind"] == "response" and "response" in event:
                    response = _dict_to_response(event["response"])
                    original_request = last_requests.pop(connection_id, None)

                    if on_response is not None:
                        result = on_response(response, original_request, ctx)
                        if asyncio.iscoroutine(result):
                            result = await result

                        if result is None:
                            await conn.pass_through(event_id)
                        else:
                            await conn.modify_response(
                                event_id, result.status,
                                result.headers, result.body,
                            )
                    else:
                        await conn.pass_through(event_id)
                else:
                    await conn.pass_through(event_id)

            except Exception:
                await conn.block(event_id)

    task = asyncio.create_task(_loop())
    return EgressInterceptHandle(task)
